cdap-ui/app/cdap/components/AbstractWidget/DatasetSelectorWidget/index.tsx

```diff
@@ -0,0 +1,239 @@
+      position: 'relative',
+    },
+    autocomplete: {
+      position: 'absolute',
+      top: 0,
+      left: 0,
+      right: 0,
+      backgroundColor: 'white',
+      border: `1px solid ${theme.palette.grey[400]}`,
+      zIndex: 1,
+      padding: '5px 0',
+      boxShadow: `0 3px 10px 0 ${theme.palette.grey[200]}`,
+    },
+    autocompleteRow: {
+      lineHeight: '1.5rem',
+      padding: '8px 10px',
+      cursor: 'pointer',
+      '&:not(:last-child)': {
+        borderBottom: `1px solid ${theme.palette.grey[400]}`,
+      },
+      '&:hover': {
+        backgroundColor: theme.palette.grey[600],
+      },
+    },
+    highlight: {
+      backgroundColor: `${theme.palette.blue[100]} !important`,
+      color: 'white',
+    },
+  };
+};
+
+interface IDatasetSelector extends WithStyles<typeof styles> {
+  onChange: (value) => void;
+  value: string;
+  placeholder?: string;
+}
+
+enum KEY_CODE {
+  Enter = 13,
+  Up = 38,
+  Down = 40,
+}
+
+export const DatasetSelectedEvent = 'dataset.selected';
+
+const DatasetSelectorView: React.SFC<IDatasetSelector> = ({
+  value,
+  onChange,
+  classes,
+  placeholder,
+}) => {
+  const [datasets, setDatasets] = React.useState([]);
+  const [showAutocomplete, setShowAutoComplete] = React.useState(false);
+  const [selection, setSelection] = React.useState(0);
+  const eventEmitter = ee(ee);
+
+  React.useEffect(() => {
+    MyDatasetApi.list({ namespace: getCurrentNamespace() }).subscribe((res) => {
+      setDatasets(res);
+    });
+  }, []);
+
+  function handleChange(e) {
+    setShowAutoComplete(true);
+    onChange(e.target.value);
+  }
+
+  function getMatch() {
+    if (!value || value.length === 0) {
+      return [];
+    }
+    return datasets.filter((dataset) => {
+      const datasetName = dataset.name.toLowerCase();
+      const inputValue = value.toLowerCase();
+      return datasetName.indexOf(inputValue) !== -1;
+    });
+  }
+
+  function handleRowClick(dataset) {
+    onChange(dataset.name);
+    const schema = objectQuery(dataset, 'properties', 'schema');
+    if (!schema) {
+      eventEmitter.emit('schema.clear');
+    } else {
+      eventEmitter.emit(DatasetSelectedEvent, schema, null, true, dataset.name);
+    }
+    setShowAutoComplete(false);
+  }
+
+  const matches = getMatch();
+
+  function changeSelection(newSelection) {
+    let finalSelection = newSelection;
+    if (newSelection < 0) {
+      finalSelection = 0;
+    } else if (newSelection > matches.length - 1) {
+      finalSelection = matches.length - 1;
+    }
+
+    setSelection(finalSelection);
+  }
+
+  function handleKeyDown(e) {
+    switch (e.nativeEvent.keyCode) {
+      case KEY_CODE.Up:
+        e.preventDefault();
+        changeSelection(selection - 1);
+        return;
+      case KEY_CODE.Down:
+        changeSelection(selection + 1);
+        return;
+    }
+  }
+
+  function handleKeyPress(e) {
+    if (e.nativeEvent.keyCode === KEY_CODE.Enter) {
+      handleRowClick(matches[selection]);
```

Comment:
This causes an issue when there are no matches. i.e dataset.name in handleRowClick where dataset is undefined.

Reply:
fixed